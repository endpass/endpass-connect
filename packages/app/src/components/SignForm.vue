<template>  
  <VFrame :loading="!request">
    <form @submit.prevent="emitSubmit">
      <FormField v-if="requesterUrl">
        <a :href="requesterUrl">{{ requesterUrl }}</a> requests sign action.
      </FormField>
      <FormField label="Requires request sign by:">
        <Message :ellipsis="true">
          {{ account }}
        </Message>
      </FormField>
      <FormField v-if="error">
        <message :error="true">{{ error }}</message>
      </FormField>
      <FormField label="Your account password:">
        <VInput 
          v-model="password" 
          :autofocus="true" 
          type="password" 
          placeholder="Enter your password..." />
      </FormField>
      <FormField 
        v-if="request" 
        label="Request data:">
        <VCode>
          {{ JSON.stringify(request.request, null, 2) }}
        </VCode>
      </FormField>
      <FormControls>
        <v-button
          :disabled="loading || !password"
          :submit="true"
          type="primary"
        >{{ primaryButtonLabel }}</v-button>
        <v-button @click="emitCancel">Close</v-button>
      </FormControls>      
    </form>
  </VFrame>
</template>

<script>
import { get } from 'lodash';
import VFrame from '@/components/VFrame.vue';
import VInput from '@/components/VInput.vue';
import VSelect from '@/components/VSelect.vue';
import VCode from '@/components/VCode.vue';
import VButton from '@/components/VButton.vue';
import Message from '@/components/Message.vue';
import FormField from '@/components/FormField.vue';
import FormControls from '@/components/FormControls.vue';

export default {
  name: 'SignForm',

  props: {
    loading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: null,
    },

    request: {
      type: Object,
      default: null,
    },

    accounts: {
      type: Array,
      default: () => [],
    },
  },

  data: () => ({
    password: '',
  }),

  computed: {
    account() {
      return get(this.request, 'address');
    },

    requesterUrl() {
      return get(this.request, 'url');
    },

    primaryButtonLabel() {
      return !this.loading ? 'Sign' : 'Loading...';
    },
  },

  methods: {
    emitSubmit() {
      this.$emit('submit', {
        account: this.account,
        password: this.password,
      });
    },

    emitCancel() {
      this.$emit('cancel');
    },
  },

  components: {
    VFrame,
    VButton,
    VInput,
    VSelect,
    VCode,
    Message,
    FormField,
    FormControls,
  },
};
</script>
