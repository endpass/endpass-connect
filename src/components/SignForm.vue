<template>  
  <form 
    data-test="sign-form" 
    @submit.prevent="emitSubmit"
  >
    <form-field v-if="requesterUrl">
      <a 
        :href="requesterUrl" 
        data-test="requester-url"
      >{{ requesterUrl }}</a> requests sign action.
    </form-field>
    <form-field label="Requires request sign by:">
      <message 
        :ellipsis="true" 
        data-test="account-address"
      >
        {{ account }}
      </message>
    </form-field>
    <form-field v-if="error">
      <message 
        :error="true" 
        data-test="error-message"
      >{{ error }}</message>
    </form-field>
    <form-field label="Your account password:">
      <v-input 
        v-model="password" 
        :autofocus="true" 
        type="password" 
        placeholder="Enter your password..."
      />
    </form-field>
    <form-field 
      v-if="requestBody" 
      label="Request data:"
    >
      <v-code data-test="request-body">
        {{ JSON.stringify(requestBody, null, 2) }}
      </v-code>
    </form-field>
    <form-controls>
      <v-button
        :disabled="loading || !password"
        :submit="true"
        type="primary"
        data-test="submit-button"
      >{{ primaryButtonLabel }}</v-button>
      <v-button 
        :disabled="!closable || loading"
        data-test="cancel-button" 
        @click="emitCancel"
      >Close</v-button>
    </form-controls>      
  </form>  
</template>

<script>
import { get } from 'lodash';
import VInput from './VInput.vue';
import VSelect from './VSelect.vue';
import VCode from './VCode.vue';
import VButton from './VButton.vue';
import Message from './Message.vue';
import FormField from './FormField.vue';
import FormControls from './FormControls.vue';

export default {
  name: 'SignForm',

  props: {
    loading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: null,
    },

    request: {
      type: Object,
      default: null,
    },

    accounts: {
      type: Array,
      default: () => [],
    },

    closable: {
      type: Boolean,
      default: true,
    },
  },

  data: () => ({
    password: '',
  }),

  computed: {
    account() {
      return get(this.request, 'address');
    },

    requesterUrl() {
      return get(this.request, 'url');
    },

    requestBody() {
      return get(this.request, 'request');
    },

    primaryButtonLabel() {
      return !this.loading ? 'Sign' : 'Loading...';
    },
  },

  methods: {
    emitSubmit() {
      if (this.password) {
        this.$emit('submit', {
          account: this.account,
          password: this.password,
        });
      }
    },

    emitCancel() {
      this.$emit('cancel');
    },
  },

  components: {
    VButton,
    VInput,
    VSelect,
    VCode,
    Message,
    FormField,
    FormControls,
  },
};
</script>
