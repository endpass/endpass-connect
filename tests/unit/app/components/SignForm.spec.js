import { shallowMount, mount } from '@vue/test-utils';
import SignForm from '@@/app/src/components/SignForm.vue';

const request = {
  address: '0x0',
  url: 'http://localhost',
  request: {
    foo: 'bar',
  },
};

describe('SignForm', () => {
  describe('render', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallowMount(SignForm, {
        propsData: {
          request,
        },
      });
    });

    it('should correctly render SignForm component', () => {
      expect(wrapper.name()).toBe('SignForm');
      expect(wrapper.find('[data-test=request-body]').exists()).toBe(true);
      expect(wrapper.find('[data-test=requester-url]').text()).toBe(
        request.url,
      );
      expect(wrapper.find('[data-test=error-message]').exists()).toBe(false);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not render form if request is not passed', () => {
      wrapper = mount(SignForm);

      expect(wrapper.find('form').exists()).toBe(false);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not render requester url if it is not passed', () => {
      wrapper = shallowMount(SignForm, {
        propsData: {
          request: {
            ...request,
            url: null,
          },
        },
      });

      expect(wrapper.find('[data-test=requester-url]').exists()).toBe(false);
    });

    it('should not render request body code if it is not passed', () => {
      wrapper = shallowMount(SignForm, {
        propsData: {
          request: {
            ...request,
            request: null,
          },
        },
      });

      expect(wrapper.find('[data-test=request-body]').exists()).toBe(false);
    });

    it('should change submit button text if loading and make it disabled', () => {
      wrapper = shallowMount(SignForm, {
        propsData: {
          loading: true,
          request,
        },
      });

      const submitButton = wrapper.find('[data-test=submit-button]');

      expect(submitButton.text()).toBe('Loading...');
      expect(submitButton.attributes().disabled).toBe('true');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(SignForm, {
        propsData: {
          request,
        },
      });
    });

    it('should not allow submit form if password is empty', () => {
      wrapper.find('form').trigger('submit');

      expect(wrapper.emitted().submit).toBe(undefined);
    });

    it('should allow submit of email is valid', () => {
      wrapper.setData({
        password: 'foo',
      });

      wrapper.find('form').trigger('submit');

      expect(wrapper.emitted().submit).toEqual([
        [
          {
            password: 'foo',
            account: request.address,
          },
        ],
      ]);
    });

    it('should cancel sign on cancel button press', () => {
      wrapper.find('[data-test=cancel-button]').trigger('click');

      expect(wrapper.emitted().cancel).toEqual([[]]);
    });
  });
});
