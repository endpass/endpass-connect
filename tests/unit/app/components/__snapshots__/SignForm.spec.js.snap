// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignForm render should change submit button text if loading and make it disabled 1`] = `
<v-frame-stub>
  <form data-test="sign-form">
    <form-field-stub><a href="http://localhost" data-test="requester-url">http://localhost</a> requests sign action.
    </form-field-stub>
    <form-field-stub label="Requires request sign by:">
      <message-stub ellipsis="true" data-test="account-address">
        0x0
      </message-stub>
    </form-field-stub>
    <!---->
    <form-field-stub label="Your account password:">
      <v-input-stub placeholder="Enter your password..." name="" value="" autofocus="true" type="password"></v-input-stub>
    </form-field-stub>
    <form-field-stub label="Request data:">
      <v-code-stub data-test="request-body">
        {
        "foo": "bar"
        }
      </v-code-stub>
    </form-field-stub>
    <form-controls-stub>
      <v-button-stub disabled="true" type="primary" submit="true" data-test="submit-button">Loading...</v-button-stub>
      <v-button-stub data-test="cancel-button">Close</v-button-stub>
    </form-controls-stub>
  </form>
</v-frame-stub>
`;

exports[`SignForm render should correctly render SignForm component 1`] = `
<v-frame-stub>
  <form data-test="sign-form">
    <form-field-stub><a href="http://localhost" data-test="requester-url">http://localhost</a> requests sign action.
    </form-field-stub>
    <form-field-stub label="Requires request sign by:">
      <message-stub ellipsis="true" data-test="account-address">
        0x0
      </message-stub>
    </form-field-stub>
    <!---->
    <form-field-stub label="Your account password:">
      <v-input-stub placeholder="Enter your password..." name="" value="" autofocus="true" type="password"></v-input-stub>
    </form-field-stub>
    <form-field-stub label="Request data:">
      <v-code-stub data-test="request-body">
        {
        "foo": "bar"
        }
      </v-code-stub>
    </form-field-stub>
    <form-controls-stub>
      <v-button-stub disabled="true" type="primary" submit="true" data-test="submit-button">Sign</v-button-stub>
      <v-button-stub data-test="cancel-button">Close</v-button-stub>
    </form-controls-stub>
  </form>
</v-frame-stub>
`;

exports[`SignForm render should not render form if request is not passed 1`] = `
<div class="frame">
  <header class="frame__header"><img src="../assets/logo.png" alt="Endpass" class="frame__logo">
    Connect
  </header>
  <div class="frame__body">
    <div class="loading-screen">
      <div class="loading-screen__spinner">
        <div class="spinner"></div>
      </div>
    </div>
  </div>
</div>
`;
